Back/kimmoCya/23

kimmoCya opened the pull request:
### Feat: 회원가입 로직 구현

Controller
- RestConrtoller를 이용(데이터를 전달하기 때문)
- RequestMapping 이용
- 입력값 올바르지 않으면 ResponseEntity.badRequest().body 반환
- 이메일 중복 시 ResponseEntity.status(HttpStatus.CONFLICT).body(e.getMessage()) 반환
- 올바르게 입력 시 ResponseEntity.ok 반환

Service
- 이메일 중복인지 아닌지 검사
- 비밀번호 해시 적용

Repository
- 이메일 중복 확인하는 코드

PusanKimC committed:
hash를 쓰기위한 gradle 추가

PusanKimC committed:
Member 컨트롤러 구현

매핑 및 회원가입 컨트롤러 구현

PusanKimC committed:
Member 서비스 구현

비밀번호 해시 적용 후 DB에 저장하게 구현.

PusanKimC committed:
Member 리포지터리 구현

이메일이 중복인지 확인하는 코드 구현.

PusanKimC committed:
Member 엔티티 Setter 추가

PusanKimC committed:
setPassword를 위한 Member 엔티티 @Setter 추가

kimmoCya commented on backend/build.gradle:
```diff
@@ -33,6 +33,7 @@ dependencies {
 	annotationProcessor 'org.projectlombok:lombok'
 	testImplementation 'org.springframework.boot:spring-boot-starter-test'
 	testRuntimeOnly 'org.junit.platform:junit-platform-launcher'
+    implementation 'org.springframework.boot:spring-boot-starter-security'
```
해시를 적용시키기 위해 org.springframework.boot:spring-boot-starter-security를 추가

tomchaccom commented on backend/src/main/java/com/example/LabAttendance/RollCall/Member/Member.java:
```diff
@@ -6,12 +6,14 @@
 import lombok.AllArgsConstructor;
 import lombok.Getter;
 import lombok.NoArgsConstructor;
+import lombok.Setter;
```
엔티티 에서는 setter를 사용하는 것이 좋지 않습니다. public으로 접근이 가능한 메소드로 객체의 값을 바꾸는 것이기 때문에 나중에 프로젝트가 커지면 어느 시점에서 setter 가 호출되었는지도 알 수없고, 캡슐화의 측면에서 고려해 보았을 때 이 객체의 내부의 정확한 구현은 모르지만, 인터페이스 즉 메소드의 인자와 리턴 값만을 알고 사용하는 것이 좋은데. setter 같은 경우는 누구나가 이 메소드의 동작을 알기 때문입니다

tomchaccom commented on backend/src/main/java/com/example/LabAttendance/RollCall/Member/Member.java:
```diff
@@ -6,12 +6,14 @@
 
 @Entity
 @Getter
+@Setter
```
그래서 setter를 사용하지 않는 2가지 방식이 있습니다. 첫째 한번만 값을 할당해도 된다면 생성자를 통해서 처리하기, 둘째 명시적인 메소드 만들기 입니다. 두번째 방식은 엔티티에게 일을 시킨다는 개념인데, 아래 서비스 코드에서 조금 더 피드백 주겠습니다

tomchaccom commented on backend/src/main/java/com/example/LabAttendance/RollCall/Member/MemberController.java:
```diff
@@ -1,13 +1,42 @@
 
 @RestController
 @RequiredArgsConstructor
+@RequestMapping("/lab_chulseokbu/user")
```
URL 에는 기본적으로 _(언더바) 를 사용하지 않는게 좋습니다. REST 설계 규칙을 한번 더 공부해보세요

tomchaccom replied:
현재 URL에서 수정할 점에 대해 짚어보자면 

1. 엔티티 또는 식별하려는 리소스(user, member) 는 URL에서는 복수형으로 나타낸다 -> users/1 이런식으로 디렉토리의 경로 관점에서 상위 디렉토리(users) 를 컬렉션으로 보기 때문
2. API 가 너무 길다 -> 캐싱의 관점에서 별로 좋지 않을 거 같습니다. lab_chulseokbu/user를 지우고 다시 설계해 보세요
3. 필터 설계 관점에서 공통된 부모 Path를 가지는 게 좋아 보입니다. 회원가입 로그인은 auth/sign, auth/login

tomchaccom commented on backend/src/main/java/com/example/LabAttendance/RollCall/Member/MemberController.java:
```diff
@@ -1,13 +1,42 @@
+    @GetMapping("/signup")
+    public String signup(Member member)
+    {
+        return "signup_form";
```
폼도 만들어서 화면으로 동작시켜 보신건가요? 그렇다면 PR에 스크린샷 같은 것도 넣어주세요

tomchaccom commented on backend/src/main/java/com/example/LabAttendance/RollCall/Member/MemberController.java:
```diff
@@ -1,13 +1,42 @@
+    @PostMapping("/signup")
+    public ResponseEntity<?> signup(@Validated @RequestBody Member member, BindingResult bindingResult) {
+        if (bindingResult.hasErrors()) {
+            return ResponseEntity.badRequest().body("입력값이 올바르지 않습니다.");
```
BindingResult는 저도 처음 보네요. 근데 저는 지금의 방식보다 Dto를 통해서 데이터를 입력받고, Dto validation을 하는게 더 좋아 보입니다. 현재 @Validated 코드에서는 어떤 필드를 어떻게 검증하겠다는 건지를 저도 저 방식으로는 코드를 안써봐서 잘 모르겠네요

tomchaccom commented on backend/src/main/java/com/example/LabAttendance/RollCall/Member/MemberService.java:
```diff
@@ -2,13 +2,35 @@
 
+        if (memberRepository.existsByEmail(member.getEmail())) {
+            throw new IllegalStateException("이미 존재하는 이메일입니다.");
+        }
```
위 조건문은 해당 이메일로 이미 DB 에 저장된 내역이 있는지를 확인하고 IllegalStateException을 던지고 있는데 이 예외는 대상 객체의 상태가 호출된 메서드를 수행하기에 적절하지 않을 때 발생시킬 수 있는 예외로 중복된 이메일을 검사하기에는 부적절 합니다. 그래서 사용자 정의 예외를 통해서 예외처리하는 것이 좋아보입니다. DuplicateUserException이런 형식으로요

tomchaccom commented on backend/src/main/java/com/example/LabAttendance/RollCall/Member/MemberService.java:
```diff
@@ -2,13 +2,35 @@
 
 @RequiredArgsConstructor
 @Service
+@Transactional
```
Transactional을 클래스에 붙일 때와 메소드에 붙일 경우의 차이에 대해서 알아보세요

tomchaccom commented on backend/src/main/java/com/example/LabAttendance/RollCall/Member/MemberService.java:
```diff
@@ -2,13 +2,35 @@
+    private final PasswordEncoder passwordEncoder;
 
+    public Long create(Member member)
+    {
```
Long 타입으로 ID를 반환하는 이유가 뭐죠? 컨트롤러에 응답에 ResponseBody가 없던데요? 
그리고 사용자가 데이터를 입력시 엔티티 타입을 모두 입력할까요? 예시를 들어 보면 앱이나 웹을 사용하면서 회원가입 시 서비스의 가입일자 같은 것을 입력한 경우는 없었을 겁니다. 그래서 이런 경우는 보통 Dto 로 입력을 받습니다.

tomchaccom reviewed (changes requested):
수정사항 반영해서 다시 PR 제출해 주세요 기한은 오늘 자정까지입니다

PusanKimC committed:
서버 구동 환경 및 기본 설정, PasswordEncoder Bean 등록

PusanKimC committed:
Member 엔티티 @Setter제거

@builder로 변수 불변성 확보
update메서드로 가변필드 변경 구현(email, phone, password)

PusanKimC committed:
회원가입 DTO

회원가입 시 서버로 보내야 하는 데이터

PusanKimC committed:
회원가입 성공 시 응답 DTO

클라이언트에 응답할 DTO
(password는 보안상 제외)

PusanKimC committed:
Member 회원가입 컨트롤러 수정

REST API url 수정, DTO를 이용한 데이터 전달

PusanKimC committed:
Member 회원가입 서비스 수정

DTO를 이용한 이메일 중복 검사
비밀번호에 해시를 적용
DB에 저장되면 프론트에 ResponseDTO 값 반환

password, email, phone 변경 미구현

PusanKimC committed:
사용자 정의 예외 정의

이메일 중복 시 예외처리 클래스

PusanKimC committed:
gradle 의존성 추가

'org.springframework.boot:spring-boot-starter-validation'

데이터 유효성 검증 추가
DTO에 검증 규칙 정의, 컨트롤러에서 검증 실행

kimmoCya commented:
<img width="2811" height="1610" alt="스크린샷 2025-11-09 164904" src="https://github.com/user-attachments/assets/0983d615-1bd1-4d66-88bf-0abfc7eb2389" />

tomchaccom commented on backend/src/main/java/com/example/LabAttendance/RollCall/Member/DTO/MemberSignupRequestDto.java:
```diff
@@ -0,0 +1,33 @@
+
+public record MemberSignupRequestDto(
+
+        @NotNull(message = "학번은 필수 입력 값입니다.")
```
@Notnull, @Notblank의 차이점을 알아보세요

tomchaccom commented on backend/src/main/java/com/example/LabAttendance/RollCall/Member/DTO/MemberSignupRequestDto.java:
```diff
@@ -0,0 +1,33 @@
+        Long memberId,
+
+        @NotBlank(message = "닉네임은 필수 입력 값입니다.")
+        String nickname,
+
+        @NotBlank(message = "비밀번호는 필수 입력 값입니다.")
+        @Size(min = 8, max = 20, message = "비밀번호는 8자 이상 20자 이하여야 합니다.")
+        String password,
+
+        @NotBlank(message = "이메일은 필수 입력 값입니다.")
+        @Email(message = "유효한 이메일 형식이 아닙니다.")
+        String email,
+
+        @NotBlank(message = "전화번호는 필수 입력 값입니다.")
+        String phone,
+
+        @NotNull(message = "성별은 필수 입력 값입니다.")
+        Gender gender
```
Dto 필드를 private로 사용하는 것이 좋아 보입니다. Dto의 값이 변경되는 시점을 내부 메소드로만 한정하기 위해서 입니다

tomchaccom commented on backend/src/main/java/com/example/LabAttendance/RollCall/Member/DuplicateEmailException.java:
```diff
@@ -0,0 +1,7 @@
+package com.example.LabAttendance.RollCall.Member;
+
+public class DuplicateEmailException extends RuntimeException {
```
사용자 예외는 global 같은 디렉토리를 추가해서 거기에 넣어주시면 됩니다

tomchaccom commented on backend/src/main/java/com/example/LabAttendance/RollCall/Member/Member.java:
```diff
@@ -38,6 +38,27 @@ public class Member {
     @Column(nullable = false)
     private Gender gender;
 
+    @Builder
```
빌더 패턴을 사용하셨네용 왜 빌더 패턴이 생성자보다 유용한지를 한번 알아보세요

tomchaccom replied:
아 그리고 클래스 상단에 @Builder 적용하면 굳이 Builder 생성자 코드 작성 안하셔도 됩니다

tomchaccom commented on backend/src/main/java/com/example/LabAttendance/RollCall/Member/MemberService.java:
```diff
@@ -1,14 +1,48 @@
+        String hashedPassword = passwordEncoder.encode(rawPassword);
+
+
+        Member member = Member.builder()
+                .memberId(requestDto.memberId())
+                .nickname(requestDto.nickname())
+                .password(hashedPassword)
+                .email(requestDto.email())
+                .phone(requestDto.phone())
+                .gender(requestDto.gender())
+                .build();
+
```
현재는 빌더를 통해서 dto 에 있는 값을 꺼내서 객체를 생성했네요. dto에 엔티티로 변환하는 로직을 추가해도 좋을 거 같아요! 그안에 빌더 패턴으로 객체를 생성하면 되거든요 메소드 형태로 모듈화하는게 좋아 보여요

tomchaccom reviewed (approved):
고생하셨습니다. 리뷰 해둔 내용을 바탕으로 기초를 좀 다져야 할것 같네요. 이제 사용자 도메인에서 필요한 기능에 대해서 생각해보고 개발해주시면 되지 싶습니다. 고생햇어요
